### ISSUE TITLE
Update passphrase encryption to use authenticated encryption

### ISSUE THREAD

**Issue by willclarktech** [LiskHQ/lisk#1819]
### Expected behavior

We should be using authenticated encryption for passphrases, as in lisk-elements.

https://github.com/LiskHQ/lisk-js/blob/development/src/cryptography/encrypt.js#L107-L124

### Actual behavior

We use AES-256-CBC, which is unauthenticated.

https://github.com/LiskHQ/lisk/blob/development/modules/delegates.js#L335

### Which version(s) does this affect? (Environment, OS, etc...)

1.0.0

**Comment by webmaster128** [comment]
What is encrypted here? Don't we have the plaintext secret in the API call and the plaintext secret in the config file anyway?

**Comment by willclarktech** [comment]
@webmaster128 Not with v1.0.0

**Comment by willclarktech** [comment]
## Summary of meeting

### Statement of the problem

In 0.9.x delegates can manage forging in one of two ways:

1. By storing plaintext passphrases in `config.json` and setting `forging.force` to `true`.
1. By sending plaintext passphrases to their node to turn forging on/off for that account.

The first option here is intended only for testing purposes, and delegates should not be using it in production. The second option raises a couple of problems:

1. Some delegates may not be using secure connections when sending API requests.
1. Some delegates may be making requests to the API using a relatively insecure server (e.g. a cheap VPS used for management).

The first problem here should be solvable with good documentation/education. The second prompts reconsideration of our current recommendation that delegates do not store plaintext passphrases in their `config.json` in production. Supposing an attacker gains root access to the delegate's forging node, there's no way to prevent theft of the passphrase. Supposing the attacker instead gains only read access to the file system, for example via a stolen backup, a level of security can be achieved by encrypting the passphrase.

In the case of a management server intended to manage automatic failover between several nodes (switching forging to a new node in case of an outage), it cannot be expected that the delegate will be able to manually input a passphrase whenever necessary.

### Proposed solution

1. Encrypt passphrases in the `config.json` on the node.
1. Store the key required for decryption wherever it's needed (e.g. on a management server).
1. Send the decryption key in the clear (but ideally using a secure connection) to the API endpoint to toggle the forging status of a node.

This way, if the decryption key is compromised, the forging status can be toggled for a while, but the delegate can just reencrypt the passphrase under a different key to regain control. If read access to the node server is gained, then the passphrase is not compromised. Both the decryption key *and* the encrypted passphrase would have to be retrieved before the passphrase would be compromised.

*Again, this proposed solution doesn't protect against all scenarios, but does raise the level of security offered against certain attack vectors.*

### Storage of encrypted secret

For storage in the `config.json`, a normal JSON object with keys for each component required to decrypt was proposed (salt, iv, tag, version, ciphertext etc), but considered too unwieldy for delegates. A more convenient option for them would be a delimited or otherwise standardised string containing all the necessary data.

In #1829 we currently have an approach attempting to align with Linux shadow files, but some problems arise with this standard:

1. It's not designed with this use case in mind.
1. It doesn't cover our specific combination of encryption procedures (PBKDF2-SHA256 and AES-256-GCM).
1. It doesn't provide support for including all our necessary parameters.
1. It uses `$` as a delimiter, which is a pain when using JavaScript (consider template strings for example).
1. The standard uses base64 but we use hex strings pretty uniformly across our products.

I propose using a simple key value format based on URL query strings. E.g.

```
version=1&salt=abcd1234&rounds=1000000&iv=abcd1234&ciphertext=abcd1234&tag=abcd1234
```

At the cost of a little space (compared to other options), we gain readability and flexibility of order. Thanks to @webmaster128 for suggesting integer versions.

### Next steps for current PR

1. Agree on string format for encrypted secrets.
1. Reencrypt relevant secrets using an IV of 12 bytes (h/t @webmaster128).
1. Store them using the new format.

**Comment by nazarhussain** [comment]
@willclarktech @MaciejBaj  @karmacoma  As per my understanding of above detail comment. So problem right now we are discussing is following: 

We have following JSON object: 

```js
{version: 1, salt: 'abcd', rounds: 1000, iv: 'abcde', ciphertext: 'abcde', tag: 'abcde'}
```
And we are deciding on any algorithm/approach to `serialize` it to `string` and `deserialize` back to `json` object. And the **challenge** we are dealing is to have that serialized string to be **more human readable**. 

The problem I see with proposed string format is it does not comply with the actual standard. We proposed that we will use `URL Query` format. Standard says that URL Query component should start with `?`, while in our case its not.  https://tools.ietf.org/html/rfc3986#section-3.4
> The query component is indicated by the first question mark ("?") character and terminated by a number sign ("#") character or by the end of the URI

The question came here in my mind, why really we need that serialization/deserialization process. Why can't we use plain JSON object in config. If required can't we use simply `JSON.stringify` and `JSON.parse`. Or we can develop completely custom serialization/deserialization algorithm for our secrets. 

What are your thoughts?

**Comment by webmaster128** [comment]
Thanks for the positive feedback @willclarktech, I appreciate.

@nazarhussain Call it `application/x-www-form-urlencoded` and you have it covered by a standard and do not require a leading `?`. This is how data in HTTP POST requests is usually sent:

```
POST / HTTP/1.1
Host: foo.com
Content-Type: application/x-www-form-urlencoded
Content-Length: 13

say=Hi&to=Mom
```

from https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST

--------------

> The question came here in my mind, why really we need that serialization/deserialization process.

Because it is much more convenient to copy a single line from an encryption tool into a config file than merging a JSON structure

**Comment by webmaster128** [comment]
I agree, it does not matter. It is a custom thing and we're defining the rules.

I would not consider `application/x-www-form-urlencoded` an URI only thing. I think of it as a general key value serialization with its origin in URLs. However, I don't have strong arguments here and this might be a matter of taste.

Looking at line 1 and 2 of your examples (I personally dislike the rest because of the unintuitive assignment character), there are pros and cons for both

**Line 1**  `version=1&salt=abcd&rounds=1000&iv=abcde&ciphertext=abcde&tag=abcde`

Pro: We have built-in support in basically every environment, e.g. node:

```
> const querystring = require('querystring');
> querystring.parse("version=1&salt=abcd&rounds=1000&iv=abcde&ciphertext=abcde&tag=abcde")
{ version: '1',
  salt: 'abcd',
  rounds: '1000',
  iv: 'abcde',
  ciphertext: 'abcde',
  tag: 'abcde' }
```

Con: you cannot use original base64 for values because `=` is part of base64 as well as the assignment character

**Line 2**  `version:1|salt:abcd|rounds:1000|iv:abcde|ciphertext:abcde|tag:abcde`

Pro: looks nice

Neutral: can still be parsed with a flexible enough library, e.g. in node:

```
> querystring.parse("version:1|salt:abcd|rounds:1000|iv:abcde|ciphertext:abcde|tag:abcde", "|", ":")
{ version: '1',
  salt: 'abcd',
  rounds: '1000',
  iv: 'abcde',
  ciphertext: 'abcde',
  tag: 'abcde' }
```

Pro: original base64 can be used for values without escaping.

Given that base64 support is not a requirement, I see both versions as equally good. However, I have a slight tendency to the first one because of tool support in different programming languages.
